src/Interpreters/Cache/FileCache.cpp

```diff
@@ -76,157 +71,104 @@ String FileCache::getPathInLocalCache(const Key & key) const
     return fs::path(cache_base_path) / key_str.substr(0, 3) / key_str;
 }
 
-void FileCache::removeKeyDirectoryIfExists(const Key & key, std::lock_guard<std::mutex> & /* cache_lock */) const
+void FileCache::assertInitialized() const
 {
-    /// Note: it is guaranteed that there is no concurrency here with files deletion
-    /// because cache key directories are create only in FileCache class under cache_lock.
-
-    auto key_str = key.toString();
-    auto key_prefix_path = fs::path(cache_base_path) / key_str.substr(0, 3);
-    auto key_path = key_prefix_path / key_str;
-
-    if (!fs::exists(key_path))
+    if (is_initialized)
         return;
 
-    fs::remove_all(key_path);
-
-    if (fs::is_empty(key_prefix_path))
-        fs::remove(key_prefix_path);
-}
-
-static bool isQueryInitialized()
-{
-    return CurrentThread::isInitialized()
-        && CurrentThread::get().getQueryContext()
-        && !CurrentThread::getQueryId().empty();
-}
-
-bool FileCache::isReadOnly()
-{
-    return !isQueryInitialized();
-}
+    std::unique_lock lock(init_mutex);
+    if (is_initialized)
```

Comment:
AFAIS, we can make `is_initialized` atomic and remove `init_mutex`

Reply:
If we want to get rid of this mutex we will also have to make `is_initialized` as not `bool` but some `enum` with states (e.g. it will be `initialization_state` instead of `is_initialized`), and I already did this some time ago, but it made code much more complicated than current version, so I'd like to leave it as is.

Reply:
Ok, I see. Then we can make `is_initialized` atomic and keep `init_mutex`.
The lock on `init_mutex` is necessary only in `initialize()`.  If `is_initialized` is atomic then there will be no need to take a mutex in assert function.

Reply:
Actually `is_initialized` is already an atomic. And `init_mutex` is needed not only in `initialize()` but also in `checkInitialized()`, because if in `checkInitialized()` we see that `is_initialized.load() == false` we should then take a mutex and check one more time because it is possible that some other thread is executing `initialize()` at this moment, but it is important for us to wait until he finishes in this case.

Reply:
Sorry, did comment w/o checking the code. Well, it depends on what contract do we want to have here. Either (1) `it's initialized or not` or (2)`it's initializing already or initialized`. Currently, we have (2).

IMO, the current way is too much, - the check is used in every public method, but `is_initialized` read-only most of the time. I wonder if we can remove this check at all. So far, I see that the cache is initialized only in `CachedObjectStorage` constructor. It means, - the cache can't be used uninitialized. Did I miss something?

Reply:
> So far, I see that the cache is initialized only in CachedObjectStorage constructor. It means, - the cache can't be used uninitialized. Did I miss something?

It can be used whenever for example AsynchronousReadIndirectBufferFromRemoteFS is used. And it can be used without CachedObjectStorage.